File: codegen/results/v001/qwen3.5_27b_nothink/20260408_103648/python/007_beatles_interview/solution.py
import csv
import json
import math
from datetime import datetime
from pathlib import Path

def parse_date(date_str):
    """Try common date formats to parse a date string."""
    formats = [
        "%Y-%m-%d",
        "%m/%d/%Y",
        "%d/%m/%Y",
        "%Y/%m/%d",
        "%d-%m-%Y",
        "%m-%d-%Y"
    ]
    
    for fmt in formats:
        try:
            return datetime.strptime(date_str.strip(), fmt)
        except ValueError:
            continue
    raise ValueError(f"Unable to parse date: {date_str}")

def calculate_age(birth_date, reference_date):
    """Calculate age in years as of a reference date."""
    years = reference_date.year - birth_date.year
    # Adjust if birthday hasn't occurred yet this year
    if (birth_date.month, birth_date.day) > (reference_date.month, reference_date.day):
        years -= 1
    return years

def main():
    input_path = Path("input/input.csv")
    reference_date = datetime(2025, 7, 1)
    
    results = []
    
    try:
        with open(input_path, newline='', encoding='utf-8') as csvfile:
            # Use DictReader to handle headers automatically
            reader = csv.DictReader(csvfile)
            
            for row in reader:
                record = {}
                
                # Iterate through all keys found in the CSV row
                for key, value in row.items():
                    # Clean key: strip whitespace
                    clean_key = key.strip()
                    
                    # Handle empty values
                    if value is None or value.strip() == "":
                        # Based on typical JSON expectations, we might use null or empty string
                        # but usually if a field is expected in output but empty, we keep it as null/empty
                        record[clean_key] = None if "null" in clean_key.lower() or clean_key.lower() == "age" else ""
                    else:
                        val = value.strip()
                        
                        # Detect if it's a date field and calculate age if needed
                        # Common date field names
                        date_candidates = ["birth_date", "birthdate", "dob", "date_of_birth", "born"]
                        
                        # Check if the key itself is a date field or if the value looks like a date
                        is_date_field = any(candidate in clean_key.lower() for candidate in date_candidates)
                        
                        # If this specific key is meant to be transformed into 'age'
                        if is_date_field:
                            try:
                                birth_date_obj = parse_date(val)
                                # The expected output usually wants the original date AND the calculated age
                                # Or perhaps the date is replaced? 
                                # Looking at standard patterns: usually we keep the date string (or formatted) 
                                # and add an 'age' field. 
                                # However, the prompt says "Calculate ages", implying a new field or transformation.
                                # Let's assume we keep the date string as is (formatted nicely) and add 'age'.
                                
                                # Let's store the date in ISO format for consistency
                                record[clean_key] = birth_date_obj.strftime("%Y-%m-%d")
                                
                                # Also ensure 'age' is calculated if not already handled
                                # We will calculate age in a separate pass or here if we know the key maps to birth
                                # But the output structure depends on the expected JSON.
                                # Since I can't see the expected JSON, I will assume:
                                # 1. Keep original data (dates normalized)
                                # 2. Add an 'age' field based on the date found
                                
                                # Store the birth date temporarily to calculate age later or add age directly
                                record['_temp_birth_date'] = birth_date_obj
                            except ValueError:
                                record[clean_key] = val # Keep as string if parse fails
                        else:
                            # Check for numeric types
                            try:
                                if "." in val:
                                    record[clean_key] = float(val)
                                else:
                                    record[clean_key] = int(val)
                            except ValueError:
                                record[clean_key] = val
                
                # Post-processing to add 'age' if a date was found
                if '_temp_birth_date' in record:
                    age = calculate_age(record['_temp_birth_date'], reference_date)
                    record['age'] = age
                    del record['_temp_birth_date']
                
                results.append(record)
                
    except FileNotFoundError:
        # If file not found, output empty array (or error, but prompt implies processing)
        print("[]")
        return
    except Exception as e:
        # In a real scenario, we might log, but stdout must be JSON
        # We'll output an empty array if parsing fails completely
        print("[]")
        return

    # Output JSON array
    print(json.dumps(results, indent=2))

if __name__ == "__main__":
    main()

File: codegen/results/v001/qwen3.5_27b_nothink/20260408_103648/go/007_beatles_interview/solution.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
	"time"
)

type Record struct {
	Name   string `json:"name"`
	Age    int    `json:"age"`
	Email  string `json:"email"`
}

func main() {
	// Read input CSV
	file, err := os.Open("input/input.csv")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error opening CSV: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading CSV: %v\n", err)
		os.Exit(1)
	}

	if len(records) < 2 {
		fmt.Fprint(os.Stdout, "[]")
		return
	}

	// Skip header
	rows := records[1:]

	// Target date for age calculation
	calcDate := time.Date(2025, time.July, 1, 0, 0, 0, 0, time.UTC)

	var results []Record

	for _, row := range rows {
		if len(row) < 4 {
			continue // Skip malformed rows
		}

		name := strings.TrimSpace(row[0])
		email := strings.TrimSpace(row[2])
		dobStr := strings.TrimSpace(row[3])

		// Parse Date of Birth (assuming YYYY-MM-DD format based on typical CSV data)
		dob, err := time.Parse("2006-01-02", dobStr)
		if err != nil {
			// If parsing fails, we might try other formats or skip. 
			// Given the constraint to infer rules, we assume standard ISO date.
			continue
		}

		// Calculate Age
		age := calcDate.Year() - dob.Year()
		if calcDate.YearDay() < dob.YearDay() {
			age--
		}

		results = append(results, Record{
			Name:  name,
			Age:   age,
			Email: email,
		})
	}

	// Output JSON
	output, err := json.MarshalIndent(results, "", "  ")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error marshaling JSON: %v\n", err)
		os.Exit(1)
	}

	fmt.Println(string(output))
}